Sources/Memcache/ChannelHandler/MemcacheBackendMessageDecoder.swift

```diff
@@ -0,0 +1,57 @@
+import Foundation
+import NIOCore
+
+struct MemcacheBackendMessageDecoder: NIOSingleStepByteToMessageDecoder {
+    typealias InboundOut = MemcacheBackendMessage
+
+    func decode(buffer: inout ByteBuffer) throws -> MemcacheBackendMessage? {
+        // Keep track of the reader index in case we later notice that we need more data
+        let startReaderIndex = buffer.readerIndex
```

Comment:
Instead of handling indexes so much, a better way is normally to get a copy of the bytebuffer.

```
var peekableBuffer = buffer

// if you were able to decode a buffer, just write the new reader indexes back:
buffer = peakableBuffer
```